src/kv/RocksDBStore.h

```diff
@@ -146,48 +152,55 @@ class RocksDBStore : public KeyValueDB {
     compact_range_async(combine_strings(prefix, start), combine_strings(prefix, end));
   }
 
-  RocksDBStore(CephContext *c, const string &path, map<string,string> opt, void *p) :
-    cct(c),
-    logger(NULL),
-    path(path),
-    kv_options(opt),
-    priv(p),
-    db(NULL),
-    env(static_cast<rocksdb::Env*>(p)),
-    dbstats(NULL),
-    compact_queue_lock("RocksDBStore::compact_thread_lock"),
-    compact_queue_stop(false),
-    compact_thread(this),
-    compact_on_mount(false),
-    disableWAL(false),
-    enable_rmrange(cct->_conf->rocksdb_enable_rmrange),
-    max_items_rmrange(cct->_conf.get_val<uint64_t>("rocksdb_max_items_rmrange"))
-  {}
-
+  RocksDBStore(CephContext *c, const string &path, map<string,string> opt, void *p);
   ~RocksDBStore() override;
 
   static bool check_omap_dir(string &omap_dir);
-  /// Opens underlying db
-  int open(ostream &out, const vector<ColumnFamily>& cfs = {}) override {
-    return do_open(out, false, false, &cfs);
-  }
+  int open(ostream &out, const std::vector<ColumnFamily>& options = {}) override;
   /// Creates underlying db if missing and opens it
-  int create_and_open(ostream &out,
-		      const vector<ColumnFamily>& cfs = {}) override;
-
-  int open_read_only(ostream &out, const vector<ColumnFamily>& cfs = {}) override {
-    return do_open(out, false, true, &cfs);
-  }
-
+  int create_and_open(ostream &out, const std::vector<ColumnFamily>& new_cfs = {}) override;
+  int open_read_only(ostream &out, const vector<ColumnFamily>& cfs = {}) override;
+  
   void close() override;
 
-  rocksdb::ColumnFamilyHandle *get_cf_handle(const std::string& cf_name) {
-    auto iter = cf_handles.find(cf_name);
-    if (iter == cf_handles.end())
-      return nullptr;
-    else
-      return static_cast<rocksdb::ColumnFamilyHandle*>(iter->second);
-  }
+  int column_family_list(vector<std::string>& cf_names) override;
+  int column_family_create(const std::string& name, const std::string& options) override;
+  int column_family_delete(const std::string& name) override;
+  KeyValueDB::ColumnFamilyHandle column_family_handle(const std::string& cf_name) const override;
+  int column_family_compact(
+		  const std::string& cf_name,
+		  const std::string& prefix,
+		  const std::string& start,
+		  const std::string& end) override;
+  int column_family_compact_async(
+		  const std::string& cf_name,
+		  const std::string& prefix,
+		  const std::string& start,
+		  const std::string& end) override;
+
+private:
+  int _open(std::ostream &out, bool read_only, const std::vector<ColumnFamily>& options);
+  std::shared_ptr<rocksdb::MergeOperator> cf_get_merge_operator(const std::string& prefix) const;
+  rocksdb::ColumnFamilyHandle* cf_get_mono_handle(const std::string& cf_name) const;
```

Comment:
would all the callers of cf_get_{mono_}handle() need to take the api_lock as well?

Reply:
@jdurgin Right, I fixed all invocations of this internal function to acquire locks.